openssl/src/pkey_ml_dsa.rs

```diff
@@ -0,0 +1,254 @@
+#[corresponds(EVP_PKEY_fromdata)]
+pub fn new_generate(variant: Variant) -> Result<PKey<Private>, ErrorStack> {
+    let mut ctx = PkeyCtx::new_from_name(None, variant.as_str(), None)?;
+    ctx.keygen_init()?;
+    let params = OsslParamBuilder::new()?.to_param()?;
+    unsafe {
+        cvt(ffi::EVP_PKEY_CTX_set_params(ctx.as_ptr(), params.as_ptr()))?;
+    }
+    ctx.generate()
+}
+
+/// Returns the Private ML-DSA PKey from the provided parameters.
+/// Creates a new `PKeyMlDsaBuilder` to generate a new ML-DSA key
+/// pair.
+#[corresponds(EVP_PKEY_fromdata)]
+pub fn new_from_seed(variant: Variant, seed: &[u8]) -> Result<PKey<Private>, ErrorStack> {
+    let mut bld = OsslParamBuilder::new()?;
+    bld.add_octet_string(OSSL_PKEY_PARAM_SEED, seed)?;
+    let mut ctx = PkeyCtx::new_from_name(None, variant.as_str(), None)?;
+    ctx.fromdata_init()?;
+    let params = bld.to_param()?;
+    unsafe {
+        let evp = cvt_p(ffi::EVP_PKEY_new())?;
+        let pkey = PKey::from_ptr(evp);
+        cvt(ffi::EVP_PKEY_fromdata(
+            ctx.as_ptr(),
+            &mut pkey.as_ptr(),
+            ffi::EVP_PKEY_KEYPAIR,
+            params.as_ptr(),
+        ))?;
+        Ok(pkey)
+    }
+}
+
+/// Returns the Private ML-DSA PKey from the provided parameters.
+/// Creates a new `PKeyMlDsaBuilder` to generate a new ML-DSA key
+/// pair.
+#[corresponds(EVP_PKEY_fromdata)]
+pub fn new_from_public(variant: Variant, public: &[u8]) -> Result<PKey<Public>, ErrorStack> {
+    let mut bld = OsslParamBuilder::new()?;
+    bld.add_octet_string(OSSL_PKEY_PARAM_PUB_KEY, public)?;
+    let mut ctx = PkeyCtx::new_from_name(None, variant.as_str(), None)?;
+    ctx.fromdata_init()?;
+    let params = bld.to_param()?;
+    unsafe {
+        let evp = cvt_p(ffi::EVP_PKEY_new())?;
+        let pkey = PKey::from_ptr(evp);
+        cvt(ffi::EVP_PKEY_fromdata(
+            ctx.as_ptr(),
+            &mut pkey.as_ptr(),
+            ffi::EVP_PKEY_PUBLIC_KEY,
+            params.as_ptr(),
+        ))?;
+        Ok(pkey)
+    }
+}
+
+pub struct PKeyMlDsaParams<T> {
+    params: OsslParamArray,
+    _m: PhantomData<T>,
+}
+
+impl<T> PKeyMlDsaParams<T> {
+    /// Creates a new `PKeyMlDsaParams` from existing ML-DSA PKey. Internal.
+    #[corresponds(EVP_PKEY_todata)]
+    fn _new_from_pkey<S>(
+        pkey: &PKey<S>,
+        selection: c_int,
+    ) -> Result<PKeyMlDsaParams<T>, ErrorStack> {
+        unsafe {
+            let mut params: *mut ffi::OSSL_PARAM = ptr::null_mut();
+            cvt(ffi::EVP_PKEY_todata(pkey.as_ptr(), selection, &mut params))?;
+            Ok(PKeyMlDsaParams::<T> {
+                params: OsslParamArray::from_ptr(params),
+                _m: PhantomData,
+            })
+        }
+    }
+}
+
+impl PKeyMlDsaParams<Public> {
+    /// Creates a new `PKeyMlDsaParams` from existing Public ML-DSA PKey.
+    #[corresponds(EVP_PKEY_todata)]
+    pub fn from_pkey<S>(pkey: &PKey<S>) -> Result<PKeyMlDsaParams<Public>, ErrorStack> {
```

Comment:
None of the pkey using functions seem to actually verify the type of pkey?

Also, this inverts the way these APIs usually work. Usually there's a PKey method to get the type-specific reprsentation.